src/rust/engine/src/scheduler.rs

```diff
@@ -76,11 +78,24 @@ impl Scheduler {
   }
 
   pub fn add_root_select(&mut self, subject: Key, product: TypeConstraint) {
+
+    // find the edges for the root in the rule graph
+    // then add a node that contains those edges to the roots
+    // TODO what to do if there isn't a match, ie if there is a root type that hasn't been specified
```

Comment:
I think the answer is that a lot of the `lib` methods should return `Result`-like types to allow for failure. But probably out of scope for now. A panic+todo for that would be fine.